### PR TITLE
Relax public API requirements on shared/lib and shared/ui

### LINKED ISSUE #76: Public API for shared/lib
The `no-public-api-sidestep` rule should consider the specifics of `shared/lib` — its public API should be one layer deeper, in `shared/lib/*/index`

### PR DESCRIPTION
**Pull request by illright**
Closes #76

### CODE REVIEW

File: packages/steiger-plugin-fsd/src/no-public-api-sidestep/index.ts [hunk 1]
```diff
@@ -64,12 +65,30 @@ const noPublicApiSidestep = {
...
 
-          const index = getIndex(segment)
-          if (resolvedDependency !== index?.path) {
-            diagnostics.push({
-              message: `Forbidden sidestep of public API when importing from "${dependency}".`,
-              location: { path: sourceFile.file.path },
-            })
+          const segmentIndex = getIndex(segment)
+          if (segment.type === 'folder' && resolvedDependency !== segmentIndex?.path) {
+            if (dependencyLocation.layerName === 'shared' && ['ui', 'lib'].includes(dependencyLocation.segmentName)) {
```
Comment (daniilsapa) [inline packages/steiger-plugin-fsd/src/no-public-api-sidestep/index.ts:70]: Non-blocking: this set of conditions is pretty difficult to read. I would create several variables that could make it possible to read it almost like plain English e.g. `isSharedLayer`, `isUIorLib` etc. What do you think

File: packages/steiger-plugin-fsd/src/public-api/index.ts [hunk 2]
```diff
@@ -15,20 +15,38 @@ const publicApi = {
...
+                    path: join(segment.path, `index.js`),
+                    // TODO: Infer better content for this file
+                    content: '',
+                  },
+                ],
+                location: { path: segment.path },
+              })
+            } else if (segment.type === 'folder') {
+              for (const child of segment.children) {
+                if (child.type === 'folder' && getIndex(child) === undefined) {
```
Comment (daniilsapa) [inline packages/steiger-plugin-fsd/src/public-api/index.ts:35]: I saw the discussion on the issue. You mentioned that it should work either with an index file for the segment or with index files in the top-level folder inside the segment. Did you decide to go just with top-level folders, or am I wrong?

Comment (daniilsapa) [reply packages/steiger-plugin-fsd/src/public-api/index.ts:35]: Also, I didn't find a check for the shared level. Should it work for all ui/lib segments throughout the codebase?

File: packages/steiger-plugin-fsd/src/no-public-api-sidestep/index.spec.ts [hunk 3]
```diff
@@ -115,8 +129,57 @@ it('reports errors on a project with a public API sidestep on shared', async ()
...
+    const root = parseIntoFsdRoot(`
+      📂 shared
+        📂 ui
+          📄 styles.ts
+          📄 Button.tsx
+          📄 TextField.tsx
+          📄 index.ts
+        📂 lib
+          📄 dates.ts
+          📂 i18n
```
[the remark this answers is not in the document]
Comment (illright) [reply packages/steiger-plugin-fsd/src/no-public-api-sidestep/index.spec.ts:149]: No, that was actually intentional. I see both shared/i18n and shared/lib/i18n as valid ways of organizing i18n related code

File: packages/steiger-plugin-fsd/src/no-public-api-sidestep/index.ts (hunk 1 above)
Comment (illright) [reply packages/steiger-plugin-fsd/src/no-public-api-sidestep/index.ts:70]: Yeah, I also thought that, but didn't know how to improve. I'll try your suggestion

File: packages/steiger-plugin-fsd/src/public-api/index.ts (hunk 2 above)
Comment (illright) [reply packages/steiger-plugin-fsd/src/public-api/index.ts:35]: > Did you decide to go just with top-level folders, or am I wrong?

That part of the code is only entered when there is no segment-level index, so both are supported.

In the issue discussion, I also mentioned that it should be one or the other, not both, but I later realized that there is no point in enforcing this

As for checking for the Shared layer, the check is implicit. We only enter that code on unsliced layers (App and Shared) and the App layer case exits early

File: packages/steiger-plugin-fsd/src/no-public-api-sidestep/index.spec.ts [hunk 4]
```diff
@@ -16,10 +16,16 @@ vi.mock('node:fs', async (importOriginal) => {
...
   return createFsMocks(
     {
+      '/shared/i18n/index.ts': '',
+      '/shared/i18n/translator.ts': '',
       '/shared/ui/styles.ts': '',
       '/shared/ui/Button.tsx': 'import styles from "./styles";',
       '/shared/ui/TextField.tsx': 'import styles from "./styles";',
       '/shared/ui/index.ts': '',
+      '/shared/lib/index.ts': '',
+      '/shared/lib/dates.ts': '',
```
[the remark this answers is not in the document]
Comment (illright) [reply packages/steiger-plugin-fsd/src/no-public-api-sidestep/index.spec.ts:26]: We already have `'shared/lib/i18n/index.ts'`, if I understand you correctly

File: packages/steiger-plugin-fsd/src/no-public-api-sidestep/index.ts (hunk 1 above)
Comment (illright) [reply packages/steiger-plugin-fsd/src/no-public-api-sidestep/index.ts]: Decided to go with this in the end:
```ts
const { layerName, segmentName } = dependencyLocation
if (layerName === 'shared' && ['ui', 'lib'].includes(segmentName)) {}
```

Comment (daniilsapa) [reply packages/steiger-plugin-fsd/src/no-public-api-sidestep/index.ts]: That works for me, thanks!

File: packages/steiger-plugin-fsd/src/public-api/index.ts (hunk 2 above)
Comment (daniilsapa) [reply packages/steiger-plugin-fsd/src/public-api/index.ts:35]: Oh, got it

Comment (daniilsapa) [reply packages/steiger-plugin-fsd/src/public-api/index.ts:35]: And yes, it looks like enforcing one or the other way is unnecessary, I agree

**Comment by daniilsapa** [comment]
Thank you for addressing the call-outs!

**Review by daniilsapa — APPROVED** [review]

**Pull request merged**